### PR TITLE
Cache url

### LINKED ISSUE #140: add api for inserting directly into the query cache
```js
store.queryURL(url, { cacheKey }).then(result => {
  store.cacheURL( computeSecondaryCacheKey(cacheKey, result), result);
})
```

unsure about name, intended to fit with [these](https://github.com/hjdivad/ember-m3/blob/master/addon/initializers/m3-store.js#L57-L67)

---

This is helpful for eg querying an seo-friendly "clean" url but still caching it for a subsequent app generated "dirty" url.

### PR DESCRIPTION
**Pull request by sangm**
Initial work for cacheURL. Fixes #140 

Few things to consider:

- What do we do with "reverseCache" _if_ anything?

### CODE REVIEW

File: addon/query-cache.js [hunk 1]
```diff
@@ -70,6 +71,12 @@ export default class QueryCache {
     }
   }
 
+  cacheURL(cacheKey, result) {
+    if (cacheKey) {
```
[the remark this answers is not in the document]
Comment (sangm) [reply addon/query-cache.js]: Assertion sounds good

File: addon/query-cache.js [hunk 2]
```diff
@@ -70,6 +72,13 @@ export default class QueryCache {
     }
   }
 
+  cacheURL(cacheKey, result) {
+    assert('cacheKey cannot be falsey', cacheKey);
```
Comment (hjdivad) [inline addon/query-cache.js]: maybe `cacheKey must be a non-empty string`?

File: tests/unit/query-cache-test.js [hunk 3]
```diff
@@ -922,4 +922,52 @@ module('unit/query-cache', function(hooks) {
...
+
+    return cachedResult.then(result => {
+      assert.deepEqual(mockResult, result, 'queryURL returned cached result');
+    });
+  });
+
+  test('.cacheURL supports unloading of models and invalidates cache', function(assert) {
+    assert.expect(2);
+
+    const firstPayload = {
```
Comment (hjdivad) [inline tests/unit/query-cache-test.js]: I prefer `let` for locally-scoped vars like this unless we're actually `Object.freeze`ing

that said I don't care much, but we should at least be consistent between `firstPayload` and `secondPayload` as neither appears to be reassigned.

File: tests/unit/query-cache-test.js [hunk 4]
```diff
@@ -922,4 +922,52 @@ module('unit/query-cache', function(hooks) {
...
+    assert.expect(2);
+    const mockResult = { id: 1, foo: 'bar' };
+
+    this.queryCache.cacheURL('foo-bar', mockResult);
+    const cachedResult = this.queryCache.queryURL('/foo-bar', { cacheKey: 'foo-bar' });
+
+    assert.ok(cachedResult && cachedResult.then, 'cachedResult is a promise');
+
+    return cachedResult.then(result => {
+      assert.deepEqual(mockResult, result, 'queryURL returned cached result');
```
Comment (hjdivad) [inline tests/unit/query-cache-test.js]: can we add an assertion that `this.adapterAjax` was not called (similar to the `.queryURL caches its results when given a cacheKey` test)

**Comment by sangm** [comment]
Fixed the issues you mentioned 👍

**Comment by hjdivad** [comment]
awesome thanks @sangm 

If you want to add something to `README.md` that would be great but otherwise no worries i'll write something up next time i release

**Pull request merged**